jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java

```diff
@@ -46,252 +45,281 @@
+
+                try
                 {
-                    if (b == '\n')
+                    while (_index < 7)
                     {
-                        _index = 7;
-                        return true;
+                        if (!parse())
+                        {
+                            if (LOG.isDebugEnabled())
+                                LOG.debug("Proxy v1 onOpen parsing ran out of bytes, marking as fillInterested");
+                            fillInterested();
+                            return;
+                        }
                     }
 
-                    LOG.warn("Bad CRLF for {}", getEndPoint());
-                    close();
-                    return false;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Proxy v1 onOpen parsing done, now upgrading");
+                    upgrade();
+                }
+                catch (Throwable x)
+                {
+                    LOG.warn("Proxy v1 error for {}", getEndPoint(), x);
+                    releaseAndClose();
                 }
             }
-            return true;
-        }
 
-        @Override
-        public void onFillable()
-        {
-            try
+            @Override
+            public ByteBuffer onUpgradeFrom()
             {
-                ByteBuffer buffer = null;
-                while (_index < 7)
-                {
-                    // Create a buffer that will not read too much data
-                    // since once read it is impossible to push back for the 
-                    // real connection to read it.
-                    int size = Math.max(1, SIZE[_index] - _builder.length());
-                    if (buffer == null || buffer.capacity() != size)
-                        buffer = BufferUtil.allocate(size);
-                    else
-                        BufferUtil.clear(buffer);
+                return _buffer;
+            }
 
-                    // Read data
-                    int fill = getEndPoint().fill(buffer);
-                    if (fill < 0)
+            @Override
+            public void onUpgradeTo(ByteBuffer prefilled)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Proxy v1 copying prefilled buffer {}", prefilled);
+                if (BufferUtil.hasContent(prefilled))
+                    BufferUtil.append(_buffer, prefilled);
+            }
+
+            /**
+             * @return true when parsing is done, false when more bytes are needed.
+             */
+            private boolean parse() throws IOException
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Proxy v1 parsing {}", BufferUtil.toDetailString(_buffer));
+                _length += _buffer.remaining();
+
+                // Parse fields
+                while (_buffer.hasRemaining())
+                {
+                    byte b = _buffer.get();
+                    if (_index < 6)
```

Comment:
again a constant? why 6 not 7 now?

Reply:
6 is the field index of the carriage return character. It used to be hardcoded, but I can make that a constant too.